### PR TITLE
[VBLOCKS-1009] feat: publish insights events

### PR DESCRIPTION
**Pull request by kpchoy**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### JIRA link(s):

- [VBLOCKS-1009](https://issues.corp.twilio.com/browse/VBLOCKS-1009)

### Description

Add insights events to callmessage sent/received/error

#### Errors
Screenshot 2024-04-09 at 11 19 07 AM

Screenshot 2024-04-08 at 4 43 33 PM

#### Sent
Screenshot 2024-04-08 at 3 53 04 PM

#### Received 
Screenshot 2024-04-08 at 3 52 49 PM

### Before review
* [ ] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [ ] Updated affected documentation
* [x] Verified locally with `npm test`
* [x] Manually sanity tested running locally
* [x] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

File: lib/twilio/call.ts [hunk 1]
```diff
@@ -1356,6 +1356,12 @@ class Call extends EventEmitter {
       messageType: messagetype,
       voiceEventSid: voiceeventsid,
     };
+    this._publisher.info('call-message', 'user-defined-message', {
```
Comment (charliesantos) [inline lib/twilio/call.ts:1359]: 'user-defined-message' shouldn't be hardcoded. let's use the `messagetype` variable. update onMessageSent as well.

File: lib/twilio/call.ts [hunk 2]
```diff
@@ -1356,6 +1356,12 @@ class Call extends EventEmitter {
...
+      call_sid: callsid,
```
Comment (charliesantos) [inline lib/twilio/call.ts:1360]: I think call_sid is auto populated from the call object. Can you try removing it and see if it still gets sent? update onMessageSent as well.

File: lib/twilio/device.ts [hunk 3]
```diff
@@ -1223,6 +1223,16 @@ class Device extends EventEmitter {
...
+      } else if (code === 31209 ) {
+        this._publisher.error('call-message', 'error', {
+          code,
+          message: customMessage,
+        }, this._activeCall);
+      } else if (code === 31100 ) {
+        this._publisher.error('call-message', 'error', {
+          code,
+          message: customMessage,
+        }, this._activeCall);
```
Comment (charliesantos) [inline lib/twilio/device.ts:1235]: If we only check for specific error codes, we will miss other future errors unless we update the code. Let's make this dynamic. I suggest use this [if block](https://github.com/twilio/twilio-voice.js/blob/feature/eventing_1.5/lib/twilio/call.ts#L1429). We already check in there if the error is related to call message. You can use the payload object to read the information you need to send to insights. Also, we need to send `voice_event_sid` as well. It's missing in your code above. See schema here https://code.hq.twilio.com/client/rtc-insights-spec/blob/master/schema/voice/event/call-message/error/index.json

**Review by charliesantos — COMMENTED** [review]
Left a review @kpchoy . Also, can you please add unit tests?

File: lib/twilio/call.ts (hunk 1 above)
Comment (kpchoy) [reply lib/twilio/call.ts:1359]: I thought we got rid of the enum for 'user-defined-message'? Should I add the enum back?

File: lib/twilio/device.ts (hunk 3 above)
Comment (kpchoy) [reply lib/twilio/device.ts:1235]: The schema has it as "not required". It looks like from the Kibana logs it is getting auto populated. Do you still want me to add it in?

File: lib/twilio/call.ts (hunk 1 above)
Comment (charliesantos) [reply lib/twilio/call.ts:1359]: No. There's a `messagetype` variable at the top. We should use that.

File: lib/twilio/device.ts (hunk 3 above)
Comment (charliesantos) [reply lib/twilio/device.ts:1235]: Are you talking about voice event sid or callsid?

Comment (kpchoy) [reply lib/twilio/device.ts:1235]: `voice_event_sid` it is not in the "required" section: https://code.hq.twilio.com/client/rtc-insights-spec/blob/master/schema/voice/event/call-message/error/index.json#L35

File: tests/unit/call.ts [hunk 4]
```diff
@@ -2085,6 +2092,33 @@ describe('Call', function() {
...
+      });
+
+      it('should publish a user-defined-message error event', () => {
+        const payloadPromise = new Promise((resolve, reject) => {
+          conn.on('error', reject);
+        });
+        payloadPromise.then(() => {
+          sinon.assert.calledWith(publisher.error, 'user-defined-message', 'error');
+        });
+      });
```
Comment (charliesantos) [inline tests/unit/call.ts:2121]: Can you also test the correct payload is published?

**Review by charliesantos — CHANGES_REQUESTED** [review]

**Review by charliesantos — APPROVED** [review]

**Pull request merged**